### PR TITLE
[ROCm] Hold the router weight in fp32 when fp32 routing is requested

### LINKED ISSUE #55351: [Bug][ROCm] gfx950 FP32 router GEMM is unreachable for DeepSeek-family models: shape (6144, 256) is listed but the weight is bf16
### Your current environment

ROCm 7.2.3, MI355X (gfx950), TP4, vLLM main + #54845.
Model: GLM-5.3 (`glm_moe_dsa`, `hidden_size=6144`, `n_routed_experts=256`, bf16).

### 🐛 Describe the bug

The gfx950 low-M FP32 router GEMM added in #54845 never runs for DeepSeek-family
models, including the `(6144, 256)` shape it explicitly supports.

`GateLinear` gates the path on `self.weight.dtype == torch.float32`, but
`deepseek_v2.py` constructs it without `params_dtype`, so the weight stays in the
model dtype (bf16):

```python
self.router_dtype = _get_moe_router_dtype(config)
self.gate = GateLinear(
    config.hidden_size,
    config.n_routed_experts,
    out_dtype=self.router_dtype,     # fp32 output ...
    prefix=f"{prefix}.gate",         # ... but the weight stays bf16
)
```

This is inconsistent with what the same function asks for:
`_get_moe_router_dtype` returns `torch.float32` unconditionally for
`model_type == "glm_moe_dsa"`, and GLM-5.3's own config carries
`"moe_router_dtype": "float32"`. The routing is declared fp32; the weight is not.

A log line at construction time makes it visible:

```
NACHWEIS Router-GEMM: gfx950=True, shape (6144, 256) supported=True,
                      weight=torch.bfloat16, bias=False -> fp32 path OFF
```

So `(6144, 256)` is in `ROCM_FP32_ROUTER_GEMM_SUPPORTED_SHAPES` and the hardware
matches, but the path is never taken. The same applies to every DeepSeek-family
model going through `deepseek_v2.py`.

### Enabling it

vLLM already has the mechanism: `GateLinear(force_fp32_compute=True)` stores the
weight in fp32 when the platform has no specialized bf16 kernel. `nemotron_h.py`
uses it; the DeepSeek/GLM path does not. Setting it where fp32 routing is already
requested flips the path on:

```python
force_fp32_compute=self.router_dtype == torch.float32,
```

```
weight=torch.float32, bias=False -> fp32 path ON
```

### Measurement

I want to be straight about the payoff: **on my setup, turning it on did not
help.** Same build, same flags, only this line differing:

| | path off | path on |
|---|---|---|
| single stream (median of 4) | 171.3 tok/s | 175.6 |
| throughput, 64 concurrent | 2005 / 2219 tok/s | 1969 / 2198 |
| GSM8K (150) | 98.0 % | 98.7 % |

That is inside run-to-run noise in both directions, and it costs ~230 MB per rank
for the fp32 copies of the router weights. So I am not proposing to flip it by
default on the strength of my numbers.

What seems worth deciding upstream is the inconsistency itself: either the
declared fp32 routing should extend to the weight (making the new kernel
reachable for the models whose shapes it lists), or the kernel's shape table
should not advertise shapes that the DeepSeek path can never reach. Right now
`(6144, 256)` is in the table and unreachable.

Happy to send whichever patch you prefer.

### PR DESCRIPTION
**Pull request by stefanskiasan**
## Purpose

Fixes #55351.

`_get_moe_router_dtype()` returns `torch.float32` unconditionally for
`glm_moe_dsa`, and GLM-5.3's own config carries `"moe_router_dtype": "float32"`.
But `GateLinear` is constructed without `params_dtype`, so only the *output* is
fp32 — the weight stays in the model dtype (bf16).

`GateLinear` gates every fp32 router GEMM on `self.weight.dtype ==
torch.float32`, so those kernels are unreachable for DeepSeek-family models. That
includes the gfx950 kernel added in #54845, whose
`ROCM_FP32_ROUTER_GEMM_SUPPORTED_SHAPES` explicitly lists `(6144, 256)` — exactly
GLM-5.3's router shape. A log line at construction time:

```
gfx950=True, shape (6144, 256) supported=True, weight=torch.bfloat16
  -> fp32 path OFF
```

vLLM already has the mechanism for this. `force_fp32_compute` stores the weight
in fp32 *only* when the platform has no specialized kernel expecting the model
dtype, so it is a no-op wherever such a kernel exists. `nemotron_h.py` uses it;
the DeepSeek path did not.

## Test Plan

MI355X (gfx950), TP4, GLM-5.3 (MXFP4 experts + fp8 attention),
`--max-model-len 1048576`, MTP k=3. Same image, same flags, only this line
differing.

## Test Result

The path is reached:

```
gfx950=True, shape (6144, 256) supported=True, weight=torch.float32
  -> fp32 path ON
```

| | path off | path on |
|---|---|---|
| single stream (median of 4) | 171.3 tok/s | 175.6 |
| throughput, 64 concurrent | 2005 / 2219 tok/s | 1969 / 2198 |
| GSM8K (150) | 98.0 % | 98.7 % |
| GPU KV cache | — | 2,843,584 tokens |

I would rather be straight about this: **on my setup it is noise in both
directions**, and it costs ~230 MB per rank for the fp32 copies of the router
weights. So I am not claiming a speedup.

What the patch fixes is the inconsistency: a model that declares fp32 routing
gets an fp32 output from a bf16 weight, and the kernels written for that case
never run. Platforms where the specialized fp32 kernel is a larger win over the
generic fallback should benefit more than mine does.

If you would rather not pay the memory for every DeepSeek-family model, the
alternative is to drop `(6144, 256)` from the ROCm shape table, since the
DeepSeek path can never reach it today. Either way the current state seems worth
resolving — happy to switch to that instead.

### CODE REVIEW

**Comment by hclsys** [comment]
this fires outside rocm too. `can_use_specialized_kernels` is `is_cuda and (is_hopper or is_blackwell)` with `is_blackwell` being family-100 only, so on sm_121 (GB10), A100 and Ada it's False — GLM-5.3 then gets an fp32 router weight, which flips off `allow_cublas_router_gemm` (gate_linear.py:127, needs a bf16 weight) and you land on the `x.to(self.weight.dtype)` upcast at :216 instead.

that tier already gives you fp32 logits: bf16 GEMM with the fp32 `out_dtype` epilogue, accumulating in fp32 (i added it in #54048, so shout if i'm wrong about your case). and the checkpoint weight is bf16 anyway, so promoting it to fp32 doubles the weight and adds an activation upcast without adding any mantissa. did you see a real accuracy diff on that path, or was the fp32 weight only needed to reach the gfx950 kernel? if it's the latter, gating on `is_gfx950 and is_rocm_fp32_shape` would leave the cuda side alone.
